pkg/crosscluster/logical/sql_row_writer.go

```diff
@@ -0,0 +1,175 @@
+// Copyright 2025 The Cockroach Authors.
+//
+// Use of this software is governed by the CockroachDB Software License
+// included in the /LICENSE file.
+
+package logical
+
+import (
+	"context"
+
+	"github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/cdcevent"
+	"github.com/cockroachdb/cockroach/pkg/sql/catalog"
+	"github.com/cockroachdb/cockroach/pkg/sql/isql"
+	"github.com/cockroachdb/cockroach/pkg/sql/parser/statements"
+	"github.com/cockroachdb/cockroach/pkg/sql/sem/tree"
+	"github.com/cockroachdb/cockroach/pkg/sql/sessiondata"
+	"github.com/cockroachdb/cockroach/pkg/sql/types"
+	"github.com/cockroachdb/errors"
+)
+
+// errStalePreviousValue is returned if the row supplied to UpdateRow,
+// UpdateTombstone, or DeleteRow does not match the value in the local
+// database.
+var errStalePreviousValue = errors.New("stale previous value")
+
+// sqlRowWriter is configured to write rows to a specific table and descriptor
+// version.
+type sqlRowWriter struct {
+	insert statements.Statement[tree.Statement]
+	update statements.Statement[tree.Statement]
+	delete statements.Statement[tree.Statement]
+
+	scratchDatums []any
+	columns       []string
+}
+
+// DeleteRow deletes a row from the table. It returns errStalePreviousValue
+// if the oldRow argument does not match the value in the local database.
+func (s *sqlRowWriter) DeleteRow(ctx context.Context, txn isql.Txn, oldRow cdcevent.Row) error {
+	s.scratchDatums = s.scratchDatums[:0]
+
+	var err error
+	s.scratchDatums, err = appendDatums(s.scratchDatums, oldRow, s.columns)
+	if err != nil {
+		return err
+	}
+
+	rowsAffected, err := txn.ExecParsed(ctx, "delete", txn.KV(),
+		sessiondata.NoSessionDataOverride,
+		s.delete,
+		s.scratchDatums...,
+	)
+	if err != nil {
+		return err
+	}
+	if rowsAffected != 1 {
+		return errStalePreviousValue
+	}
+	return nil
+}
+
+// InsertRow inserts a row into the table. It will return an error if the row
+// already exists.
+func (s *sqlRowWriter) InsertRow(ctx context.Context, txn isql.Txn, row cdcevent.Row) error {
+	s.scratchDatums = s.scratchDatums[:0]
+
+	var err error
+	s.scratchDatums, err = appendDatums(s.scratchDatums, row, s.columns)
+	if err != nil {
+		return err
+	}
+
+	_, err = txn.ExecParsed(ctx, "insert", txn.KV(),
+		sessiondata.NoSessionDataOverride,
+		s.insert,
+		s.scratchDatums...,
+	)
+	return err
```

Comment:
Would it be worth adding an assertion that rowsAffected is 1 after a successful insert?